samples/core/multi-device/main.cpp

```diff
@@ -0,0 +1,501 @@
+        const size_t half_input_bytes =
+            sizeof(cl_float) * pad_x_dim * (pad_grid_midpoint + 1);
+        const size_t input_offset =
+            sizeof(cl_float) * pad_x_dim * (pad_grid_midpoint - 1);
+        const size_t half_output_bytes =
+            sizeof(cl_float) * x_dim * grid_midpoint;
+
+        std::vector<
+            cl::KernelFunctor<cl::Buffer, cl::Buffer, cl::Buffer, cl_uint2>>
+            convolutions{};
+        std::vector<cl::CommandQueue> sub_queues{};
+        std::vector<cl::Buffer> sub_input_grids{}, sub_output_grids{};
+
+        for (size_t i = 0; i < subdevices.size(); ++i)
+        {
+            auto subdevice = subdevices[i];
+
+            if (diag_opts.verbose)
+            {
+                std::cout
+                    << "\n  Creating kernel and command queue of sub-device "
+                    << i << "...";
+                std::cout.flush();
+            }
+
+            auto convolution =
+                cl::KernelFunctor<cl::Buffer, cl::Buffer, cl::Buffer, cl_uint2>(
+                    program, "convolution_3x3")
+                    .getKernel();
+
+            cl::CommandQueue queue(context, subdevice,
+                                   cl::QueueProperties::Profiling);
+
+            // Initialize device-side storage.
+            // First device performs the convolution in the upper half and
+            // second device in the lower half (middle borders included).
+            if (diag_opts.verbose)
+            {
+                std::cout << "done.\n  Initializing device-side storage of "
+                             "sub-device "
+                          << i << "...";
+                std::cout.flush();
+            }
+
+            cl_buffer_region input_region = { i * input_offset,
+                                              half_input_bytes },
+                             output_region = { i * half_output_bytes,
+                                               half_output_bytes };
+
+            const cl_uint align =
+                subdevice.getInfo<CL_DEVICE_MEM_BASE_ADDR_ALIGN>();
+            if (input_region.origin % align || output_region.origin % align)
+            {
+                std::cerr << "Error: Memory should be aligned to "
+                          << subdevice.getInfo<CL_DEVICE_MEM_BASE_ADDR_ALIGN>()
+                          << std::endl;
+                exit(EXIT_FAILURE);
+            }
+
+            cl::Buffer sub_input_grid = dev_input_grid.createSubBuffer(
+                CL_MEM_READ_ONLY, CL_BUFFER_CREATE_TYPE_REGION, &input_region);
+            cl::Buffer sub_output_grid = dev_output_grid.createSubBuffer(
+                CL_MEM_WRITE_ONLY, CL_BUFFER_CREATE_TYPE_REGION,
+                &output_region);
+
+            if (diag_opts.verbose)
+            {
+                std::cout << "done.";
+                std::cout.flush();
+            }
+
+            convolutions.push_back(convolution.clone());
```

Comment:
This is correct, but the C version creates a new kernel handle whereas this C++ version clones the kernel.  It would be nice for both versions to stay consistent.

Actually, it looks like this C++ sample creates a kernel and then clones it in the loop.  Is this needed?  Or can we just use the created kernel handle without cloning it?

Reply:
The reason for cloning it is that copying a kernel handle actually does a shallow copy, so setting an arg of one kernel also modifies the other.

Reply:
Right, but we're not copying a kernel handle, we're creating a new one.  With the latest commit, this is happening in the loop:

```c++
auto convolution = cl::Kernel(program, "convolution_3x3");
```

Reply:
Yep, you're right. Should be changed now.